java/ql/src/experimental/Security/CWE/CWE-094/JythonInjection.ql

```diff
@@ -0,0 +1,120 @@
+    getName().matches(["exec%", "run%", "eval", "compile"])
+  }
+}
+
+/** The class `org.python.core.BytecodeLoader`. */
+class BytecodeLoader extends RefType {
+  BytecodeLoader() { this.hasQualifiedName("org.python.core", "BytecodeLoader") }
+}
+
+/** Holds if a Jython expression if evaluated, compiled or executed. */
+predicate runCode(MethodAccess ma, Expr sink) {
+  exists(Method m | m = ma.getMethod() |
+    m instanceof InterpretExprMethod and
+    sink = ma.getArgument(0)
+  )
+}
+
+/** A method that loads Java class data. */
+class LoadClassMethod extends Method {
+  LoadClassMethod() {
+    this.getDeclaringType().getAnAncestor*() instanceof BytecodeLoader and
+    hasName(["makeClass", "makeCode"])
+  }
+}
+
+/**
+ * Holds if `ma` is a call to a class-loading method, and `sink` is the byte array
+ * representing the class to be loaded.
+ */
+predicate loadClass(MethodAccess ma, Expr sink) {
+  exists(Method m, int i | m = ma.getMethod() |
+    m instanceof LoadClassMethod and
+    m.getParameter(i).getType() instanceof Array and // makeClass(java.lang.String name, byte[] data, ...)
+    sink = ma.getArgument(i)
+  )
+}
+
+/** The class `org.python.core.Py`. */
+class Py extends RefType {
+  Py() { this.hasQualifiedName("org.python.core", "Py") }
+}
+
+/** A method declared on class `Py` or one of its descendants that compiles Python code. */
+class PyCompileMethod extends Method {
+  PyCompileMethod() {
+    this.getDeclaringType().getAnAncestor*() instanceof Py and
+    getName().matches("compile%")
+  }
+}
+
+/** Holds if source code is compiled with `PyCompileMethod`. */
+predicate compile(MethodAccess ma, Expr sink) {
+  exists(Method m | m = ma.getMethod() |
+    m instanceof PyCompileMethod and
+    sink = ma.getArgument(0)
+  )
+}
+
+/** An expression loaded by Jython. */
+class CodeInjectionSink extends DataFlow::ExprNode {
+  CodeInjectionSink() {
+    runCode(_, this.getExpr()) or
+    loadClass(_, this.getExpr()) or
+    compile(_, this.getExpr())
+  }
+
+  MethodAccess getMethodAccess() {
+    runCode(result, this.getExpr()) or
+    loadClass(result, this.getExpr()) or
+    compile(result, this.getExpr())
+  }
+}
+
+class CodeInjectionConfiguration extends TaintTracking::Configuration {
+  CodeInjectionConfiguration() { this = "CodeInjectionConfiguration" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof CodeInjectionSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    // @RequestBody MyQueryObj query; interpreter.exec(query.getInterpreterCode());
+    exists(MethodAccess ma |
+      ma.getMethod().getDeclaringType().getASubtype*() instanceof SpringUntrustedDataType and
+      not ma.getMethod().getDeclaringType() instanceof TypeObject and
+      ma.getQualifier() = node1.asExpr() and
+      ma = node2.asExpr()
+    )
+  }
+}
```

Comment:
This is not Jython specific and should be moved to the right Spring lib

Reply:
Thanks @pwntester for reviewing this PR. As I mentioned above, in the [sample project](https://github.com/nejeoui/interpreter) I referenced,  the Getter method <code>getInterpreterCode</code> doesn't have an instance attribute backed thus is not evaluated as a <code>GetterMethod</code> therefore not detected by the CodeQL library. I doubt we shall change <code>TaintTrackingUtil.qll</code> to accommodate this scenario.

As this is an edge case, I would recommend to remove the <code> isAdditionalTaintStep</code> check to reduce the confusion unless you or Chris disagrees. Although usage in the sample project won't be detected, the query can detect most usage in GitHub repositories and the query has a clearer logic.

Reply:
I think this is similar to the case discussed in the second LiveQL episode where we want to enable taintsteps from `this` to any field accesses for RemoteFlowSource defined as Parameters. In the live stream, we did something along the lines of:

```
RefType entrypointType() {
  result = any(RemoteFlowSource s).getType() or
  result = entrypointType().getASubtype() or
  result = entrypointType().getAField().getType()
}

class EntrypointFieldStep extends TaintTracking::AdditionalTaintStep {
  override predicate step(DataFlow::Node a, DataFlow::Node b) {
    a = DataFlow::getFieldQualifier(b.asExpr().(FieldRead)) and
    a.getType() = entrypointType()
  }
}
```

@smowton Would it be possible to narrow `entrypointType` down to only those RemoteFlowSources which are defined using `asParameter()` (eg: parameters of Spring MVC / Spring REST / Jersey controller methods) ? Perhaps:

```
RefType entrypointType() {
  result = any(RemoteFlowSource s | exists(Parameter p | p = s.asParameter())).getType() or
  result = entrypointType().getASubtype() or
  result = entrypointType().getAField().getType()
}
```

@luchua-bc Please remove it the additional step from the query. we will take into consideration that False Negative

Reply:
Thanks @pwntester for the detailed explanation. I've removed it from the query.